### ISSUE TITLE
crash in the cairo library (CanvasRenderingContext2d.cc)

### ISSUE THREAD

**Issue by gravsten** [Automattic/node-canvas#790]
If distortion is enabled (default), there is a crash in the cairo library:
CanvasRenderingContext2d.cc
-> NAN_METHOD(Context2d::SetFont) {
-> cairo_set_font_size(ctx, size);

The problem always occurs at sample #7 and character #19 ('j'); config.json is attached to this bug report.

I confirmed this problem with canvas version 1.4.0 and 1.3.17; either work fine if distortion is disabled.

The cairo library is from gtk+-bundle_2.22.1-20101229_win64.zip
(tested on a Windows 8.1 64-bit system),
with GTK_Root=C:\Progra~1\GTK2

Anyone has encountered this problem before ?
Obviously without distortion the training isn't helpful to solve captchas.

[config.txt](https://github.com/Automattic/node-canvas/files/351849/config.txt)

**Comment by LinusU** [comment]
Eh, what is the config.json?

Can you provide a minimum test case using just Node.js and the canvas lib that fails?

**Comment by gravsten** [comment]
Sorry, to be more precise, I have just installed OCR (https://github.com/mateogianolio/ocr) that depends on node-canvas. Please find below detailed information on how to replicate my setup:
[HOWTO.txt](https://github.com/Automattic/node-canvas/files/351918/HOWTO.txt)

OCR's config.json is in attachment to my original message (had to rename it "config.txt" to conform to GitHub's file extension police).

In that config.json, you will find:
"distortion": false

There, just run in OCR's root folder:
node main.js

This will work fine (distortion is false).

Change config.json to:
"distortion": true

Then again run in OCR's root folder:
node main.js

And this time it will fail within seconds.

How do you suggest that I create a minimum test case using just Node.js and the canvas lib ?

**Comment by LinusU** [comment]
Looking at that library it would probably look something like this

```js
const Canvas = require('canvas')
const canvas = new Canvas(100, 100)
const context = canvas.getContext('2d')

const config = /* ... */

context.fillStyle = config.background
context.fillRect(0, 0, config.width, config.height)
context.fillStyle = config.color

context.setTransform(
  Math.random() * 0.25 + 1, // scale horizontally
  Math.random() * 0.25, // skew horizontally
  Math.random() * 0.25, // skew vertically
  Math.random() * 0.25 + 1, // scale vertically
  config.height * i + (config.height - size) / 2, // move horizontally
  config.height - size / 2 // move vertically
)

context.fillText(config.text.charAt(i), 0, 0)
```

**Comment by gravsten** [comment]
You are right, the only difference between distortion enabled / disabled is the first 4 parameters given to context.setTransform(). Without distortion, it is simply:

        context.setTransform(
          1, // scale horizontally
          0, // skew horizontally
          0, // skew vertically
          1, // scale vertically
          0, // move horizontally
          0 // move vertically
        );

which leads us to node-canvas/context2d.js where setTransform() is defined:
Context2d.prototype.setTransform = function(){
  this.resetTransform();
  this.transform.apply(this, arguments);
};

In my tests, the crash always occurs on the 272th time (after run 0..6 of 36 characters, and after the first characters 0..18) that setTransform() and fillText() are called.

I will try to come up with the proper minimum test case and come back to you shortly.

**Comment by gravsten** [comment]
There, I have made a proper minimum test case:
[bug.txt](https://github.com/Automattic/node-canvas/files/352027/bug.txt)

Just rename bug.txt to bug.js, and run:
node bug.js

This will crash within seconds.

**Comment by gravsten** [comment]
Hi,

Have you been able to confirm that there is a problem with the minimum test case I uploaded ?

If you have access to a Linux or OSX box, it would help to see if the same problem affects those other systems too -> if not, the problem probably lies in the Cairo library, requiring us to update to a newer version (GTK2's latest Windows binaries are already 5 years old!) before considering to contact the Cairo mailing list.

**Comment by zbjornson** [comment]
Your example doesn't crash for me -- it gets through all of the iterations successfully.

gtk+-bundle_2.22.1-20101229_win64
node v6.2.2
Win 7 x64
canvas 1.4.0

**Comment by LinusU** [comment]
The example works for me as well.

OS X 10.11.4
Node.js v6.2.1
Canvas 1.4.0

**Comment by gravsten** [comment]
gtk+-bundle_2.22.1-20101229_win64.zip
node v.6.2.1
Win 8.1 x64
canvas 1.4.0

And it still crashes,
even after I reinstalled all dependency modules and then rebuilt again:
npm rebuild --GTK_Root=C:\Progra~1\GTK2 --jpeg_root=C:\Progra~1\libjpeg-turbo

  Canvas.cc
  CanvasGradient.cc
  CanvasPattern.cc
  CanvasRenderingContext2d.cc
  color.cc
  Image.cc
  ImageData.cc
  init.cc
  win_delay_load_hook.c
C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\src\win_delay_load_hook.c(31): warning C4055: 'type cast' : from data pointer 'HMODULE' to function pointer 'FARPROC' [C:\Wamp\www\ocr\node_modules\canvas\build\canvas.vcxproj]
     Creating library C:\Wamp\www\ocr\node_modules\canvas\build\Release\canvas.lib and object C:\Wamp\www\ocr\node_modules\canvas\build\Release\canvas.exp
  Generating code
  Finished generating code
  canvas.vcxproj -> C:\Wamp\www\ocr\node_modules\canvas\build\Release\\canvas.node
  Copying C:\Progra~1\GTK2/bin/libexpat-1.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\libexpat-1.dll
          1 fichier(s) copié(s).
  Copying C:\Progra~1\GTK2/bin/libpng14-14.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\libpng14-14.dll
          1 fichier(s) copié(s).
  Copying C:\Progra~1\GTK2/bin/libfreetype-6.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\libfreetype-6.dll
          1 fichier(s) copié(s).
  Copying C:\Progra~1\GTK2/bin/zlib1.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\zlib1.dll
          1 fichier(s) copié(s).
  Copying C:\Progra~1\GTK2/bin/libcairo-2.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\libcairo-2.dll
          1 fichier(s) copié(s).
  Copying C:\Progra~1\GTK2/bin/libfontconfig-1.dll to C:\Wamp\www\ocr\node_modules\canvas\build\Release\\libfontconfig-1.dll
          1 fichier(s) copié(s).
  win_delay_load_hook.c
  Generating code
  Finished generating code
  canvas-postbuild.vcxproj -> C:\Wamp\www\ocr\node_modules\canvas\build\Release\\canvas-postbuild.node
canvas@1.4.0 C:\Wamp\www\ocr\node_modules\canvas
nan@2.4.0 C:\Wamp\www\ocr\node_modules\nan
pngjs@2.3.1 C:\Wamp\www\ocr\node_modules\pngjs
synaptic@1.0.7 C:\Wamp\www\ocr\node_modules\synaptic

node bug.js

0 of 3000
  0: 0
  1: 1
...
7 of 3000
  0: 0
  1: 1
  2: 2
  3: 3
  4: 4
  5: 5
  6: 6
  7: 7
  8: 8
  9: 9
  10: a
  11: b
  12: c
  13: d
  14: e
  15: f
  16: g
  17: h
  18: i
  19: j

**Comment by gravsten** [comment]
Any idea of a good tool or debug environment to trace any exception or other problem in node-canvas and its dependencies ?
or how to enable debug outputs from the Cairo canvas ?

**Comment by zbjornson** [comment]
If you have visual studio installed, then try this:

1. <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>O</kbd> and open build/binding.sln.
2. Debug menu > "canvas Properties..." > "Debugging" section: change "Command" to "C:/your/path/to/node"
3. <kbd>F5</kbd> to start debugging. You should get a node.exe command prompt popping up.
4. In the prompt, `require("bug.js")` to load your file. Hopefully VS will break on the exception. If it doesn't break, then check out the docs [here](https://msdn.microsoft.com/en-us/library/x85tt0dd.aspx) for enabling a broader range of exception types -- you can just enable them all if you're willing to deal with a few spurious messages, potentially.

**Comment by gravsten** [comment]
Thanks to your detailed explanation, I managed to get more info:

Unhandled exception at 0x0000000068DF85D2 (libcairo-2.dll) in node.exe: 0xC0000005: Access violation reading location 0x000000003F9789B8.

This is a Release version of Cairo, so there is no more than that.
I will see tomorrow if there is any way to install a Debug version of Cairo, or even better build Cairo from source.

**Comment by zbjornson** [comment]
FWIW... just re-visited upgrading to GTK3/cairo 1.14, and was reminded that it's a miserable mess to try to build cairo. The precompiled distributions don't have libpng.lib, and I'm not sure if building from source would create them.

**Comment by chearon** [comment]
FWIW I think it's fairly easy to build Cairo in MSYS2, it's a port of GCC, cmake, etc. to native Windows. Actually it's now the official way to get GTK+ and Cairo

**Comment by gravsten** [comment]
I have managed to build the latest version of Cairo on my system (Windows 8.1 x64), as one single DLL (all dependencies inside, both zlib and libpng).

The build system of node-canvas expects GTK2 and includes several custom build commands referring to other DLLS:
libexpat-1.dll
libfontconfig-1.dll
libfreetype-6.dll
libpng14-14.dll
zlib1.dll

Am I right to believe these are just dependencies of GTK2's libcairo-2.dll (i.e. node-canvas does not need them directly)?

**Comment by chearon** [comment]
Yeah, canvas might use libpng, but the windows build doesn't use FT or any of those other DLLs directly. You can see what the Cairo DLL uses with [DLL explorer](http://www.dependencywalker.com). Canvas just manually copies all of Cairo's dependencies (and their dependencies) into the release folder

**Comment by gravsten** [comment]
I have found the solution.

After building Cairo from source (and this ain't easy), the problem was the same as described above.
So I tinkered a bit more with node-canvas and the minimum test case, to realize that the problem was a simple exhaustion of some (platform specific) resource.

Indeed, the minimum test case creates a new canvas and context at each loop iteration.
By taking this outside of the loop, it works !
